### ISSUE TITLE
Engine - Introducing Events per Second (EPS) Limiting Mechanism in Wazuh Engine

### ISSUE THREAD

**Issue by JcabreraC** [wazuh/wazuh#18991]
| Wazuh version | Component   | Install type | Install method | Platform |
|---------------|-------------|--------------|----------------|----------|
| 5.0           | Wazuh Engine| Manager      | Packages/Sources| any      |

### Description
At present, the Wazuh Engine processes all incoming events at the maximum achievable speed without any form of rate limiting in place.

The purpose of this issue is to propose the incorporation of an Events per Second (EPS) Limiting Mechanism into the Wazuh Engine. This mechanism is essential for enforcing a limit on the number of events processed per second within the new engine. 

The limiting mechanism should perform the same as the actual one, meaning that events should be enqueued when above the limit, discarding the events only when the queue is full.

### Background
To align with Wazuh's pricing model, it is necessary to perform a comprehensive study of the implementation of this EPS limiting mechanism. This feature will play a crucial role in ensuring optimal resource utilization and maintaining cost-effectiveness for our users.

### Tasks
- [x] Implement EPS limiter
- [x] Update API to change limiter settings in runtime
- [x] Update CLI commands to change limiter settings in runtime
- [x] Update and add new tests
- [x] Update documentation

**The modification of the queue module to discard events instead of saving them on the flood file, will be done on a separate issue**

**Comment by JavierBejMen** [comment]
## Analysis
###Current Analysisd Implementation
Analysisd uses a windowed counter of events processed. As long as the maximum threshold is not reached, it continues to pull events from the queue to process them. This includes all types of events, such as WazuhDB synchronization and other internal events. When the maximum threshold is reached, additional events remain in the general queue and are dropped if the queue becomes full.

This method allows for burst ingestion without loss, provided the mid-term throughput stays below the threshold.

However, this approach introduces a potential issue: under conditions of saturation, internal events crucial for Wazuh DB synchronization may be deferred or lost.

### Engine
#### Proposal 1
In the Engine, we could replicate the existing behavior by limiting the number of events that processing threads pull from the router's queue. However, the issue with internal events remains.

To address this, creating a separate queue for internal events could be a solution.

#### Proposal 2
If we prefer not to alter the router and wish to implement rate limiting solely in the server module to mimic the current behavior, another queue will need to be introduced within the server.

This new queue would only handle events originating from logs. Internal events would always be directed to the router, necessitating the implementation of this filtering mechanism.

This setup would monitor the rate of events ingested per second. To account for processed events, the router's threads would need to update the counter in any scenario.

**Comment by JavierBejMen** [comment]
## Update
Finally, we are going to implement the first proposal, maintaining the actual behavior.

**Comment by JavierBejMen** [comment]
## Update
The EpsCounter class within the base namespace is designed to manage a lock free counter with a rate-limiting feature, allowing it to be safely incremented across multiple threads while ensuring the counter is reset only once per predefined interval. The class provides a simple interface with an `update()` method to increment the counter and a `limitReached()` method to check if the limit has been exceeded and reset the counter if necessary.

Class Design and Interface:
- Initialization: Configured with a limit and an interval in seconds upon creation, automatically adjusting the interval to nanosecond precision.
- Update Method: Safely increments the counter with minimal overhead using `std::memory_order_relaxed`.
Limit Check: Evaluates whether the counter has reached the predefined limit and manages the interval-based reset.

Key Implementation Points:
- Lock free, using atomics.
- Nanosecond Precision: Ensures alignment with `std::chrono::steady_clock` precision for accurate interval timing.
- Efficient Counter Updates: Utilizes relaxed memory ordering for fast and atomic counter increments.
- Synchronized Reset: Employs a strong compare-and-exchange mechanism with acquire-release semantics to guarantee a single reset operation per interval, ensuring thread-safe updates.
- Optimized Limit Check: Incorporates a preemptive check to avoid unnecessary computations, enhancing performance.

Example usage:
```c++
// Process production queue
base::Event event {};
if (!epsCounter->limitReached() && m_rQueue->waitPop(event, WAIT_DEQUEUE_TIMEOUT_USEC) && event != nullptr)
{
    m_router->ingest(std::move(event));
    if (epsCounter)
    {
        epsCounter->update();
     }
}
```

**Comment by JavierBejMen** [comment]
## Update
Moved EpsCounter class to private code on orchestrator (PIMP) and added api methods to change, get and de/activate it.

**Comment by JavierBejMen** [comment]
## Update
Added component and unit tests.
Added atomic variables for EPS settings to be able to modify them on runtime through the API
